File: CallkitSample/CallkitSample/StringeeImplement/StringeeImplement.m
//
//  StringeeImplement.m
//  SampleVoiceCall
//
//  Created by Hoang Duoc on 10/25/17.
//  Copyright © 2017 Hoang Duoc. All rights reserved.
//

#import "StringeeImplement.h"
#import "CallManager.h"
#import "CallingViewController.h"
#import "InstanceManager.h"

@implementation StringeeImplement {
    NSString *_deviceToken;
    BOOL _registeredPush;
}

static StringeeImplement *sharedMyManager = nil;

+ (StringeeImplement *)instance {
    @synchronized(self) {
        if (sharedMyManager == nil) {
            sharedMyManager = [[self alloc] init];
        }
    }
    return sharedMyManager;
}

- (id)init {
    self = [super init];
    if (self) {
        // Khởi tạo StringeeClient
        self.stringeeClient = [[StringeeClient alloc] initWithConnectionDelegate:self];
        self.stringeeClient.incomingCallDelegate = self;
    }
    return self;
}

// Kết nối tới stringee server
-(void) connectToStringeeServer {
    [self.stringeeClient connectWithAccessToken:@"YOUR_ACCESS_TOKEN"];
}

- (void)registerPushTokenIfNeedWithToken:(NSString *)deviceToken {
    if (deviceToken != nil || deviceToken.length > 0) {
        _deviceToken = deviceToken;
    }
    
    if (_deviceToken == nil || _deviceToken.length == 0 || _registeredPush || !self.stringeeClient.hasConnected) {
        return;
    }
        
    // Note: remember to pass isProduction depends on environment you are working on (development or production)
    [self.stringeeClient registerPushForDeviceToken:_deviceToken isProduction:false isVoip:true completionHandler:^(BOOL status, int code, NSString *message) {
        NSLog(@"registerPush: %@", message);
        _registeredPush = status;
    }];
}

// MARK:- Stringee Connection Delegate

// Lấy access token mới và kết nối lại đến server khi mà token cũ không có hiệu lực
- (void)requestAccessToken:(StringeeClient *)StringeeClient {
    [self.stringeeClient connectWithAccessToken:@"YOUR_ACCESS_TOKEN"];
}

- (void)didConnect:(StringeeClient *)stringeeClient isReconnecting:(BOOL)isReconnecting {
    NSLog(@"didConnect");
    dispatch_async(dispatch_get_main_queue(), ^{
        if ([InstanceManager instance].homeVC) {
            [InstanceManager instance].homeVC.title = stringeeClient.userId;
            [InstanceManager instance].homeVC.btCall.enabled = YES;
            [InstanceManager instance].homeVC.btVideoCall.enabled = YES;
        }
        
        [self registerPushTokenIfNeedWithToken:nil];
    });
}

- (void)didDisConnect:(StringeeClient *)stringeeClient isReconnecting:(BOOL)isReconnecting {
    NSLog(@"didDisConnect");
    dispatch_async(dispatch_get_main_queue(), ^{
        [InstanceManager instance].homeVC.btCall.enabled = NO;
    });
}

- (void)didFailWithError:(StringeeClient *)stringeeClient code:(int)code message:(NSString *)message {
    NSLog(@"didFailWithError - %@", message);
}

- (void)incomingCallWithStringeeClient:(StringeeClient *)stringeeClient stringeeCall:(StringeeCall *)stringeeCall {
    [[CallManager instance] handleIncomingCallEvent:stringeeCall];
}

// MARK: - Private Method

- (void)delayCallback:(void(^)(void))callback forTotalSeconds:(double)delayInSeconds {
    dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, delayInSeconds * NSEC_PER_SEC);
    dispatch_after(popTime, dispatch_get_main_queue(), ^(void){
        if(callback){
            callback();
        }
    });
}

@end

File: CallkitSample-iOS13/CallkitSample-iOS13/StringeeImplement/StringeeImplement.swift
/*
 - Kết nối tới Stringee Server
 - Xử lý các sự kiện liên quan đến kết nối
 */

import UIKit

let userToken1 = "eyJjdHkiOiJzdHJpbmdlZS1hcGk7dj0xIiwidHlwIjoiSldUIiwiYWxnIjoiSFMyNTYifQ.eyJqdGkiOiJTSy4wLjVzNlBlYjlZcXpaVDRQWWoxNG1Gajhwa0hMd2dLLTE2Nzk1NjI5ODMiLCJpc3MiOiJTSy4wLjVzNlBlYjlZcXpaVDRQWWoxNG1Gajhwa0hMd2dLIiwiZXhwIjoxNjgyMTU0OTgzLCJ1c2VySWQiOiJ1c2VyMSJ9.p8ZgfYZ-deI2-F-VMzdDJRjy1i4kPLt44OZEbNPJ4eg"
let userToken2 = "eyJjdHkiOiJzdHJpbmdlZS1hcGk7dj0xIiwidHlwIjoiSldUIiwiYWxnIjoiSFMyNTYifQ.eyJqdGkiOiJTSy4wLjVzNlBlYjlZcXpaVDRQWWoxNG1Gajhwa0hMd2dLLTE2Nzk1NjI5OTMiLCJpc3MiOiJTSy4wLjVzNlBlYjlZcXpaVDRQWWoxNG1Gajhwa0hMd2dLIiwiZXhwIjoxNjgyMTU0OTkzLCJ1c2VySWQiOiJ1c2VyMiJ9.LDQzSsz2M31DUdQbCFfFXiK-fjt5zqng2wZrgEVdJOo"

class StringeeImplement: NSObject {
    
    static let shared = StringeeImplement()
    let stringeeClient = StringeeClient()
    
    var token: String?
    var connectWithToken1 = false
    
    // push
    var pushToken: String?
    var registeredTokenForPush = false
    
    private override init() {
        super.init()
        stringeeClient.connectionDelegate = self;
        stringeeClient.incomingCallDelegate = self
        token = connectWithToken1 ? userToken1 : userToken2
    }
    
    func connectToStringeeServer() {
        stringeeClient.connect(withAccessToken: token)
    }
    
    func registerTokenForPush(token: String? = "") {
        self.pushToken = token
        guard let pushToken = self.pushToken, !pushToken.isEmpty, !registeredTokenForPush else {
            return
        }
        
        // Note: remember to pass isProduction depends on environment you are working on (development or production)
        self.stringeeClient.registerPush(forDeviceToken: pushToken, isProduction: false, isVoip: true) { (status, code, message) in
            print("registerPush: \(String(describing: message))")
            self.registeredTokenForPush = status
        }
    }
}

extension StringeeImplement: StringeeConnectionDelegate {
    
    func didConnect(_ stringeeClient: StringeeClient!, isReconnecting: Bool) {
        print("didConnect \(stringeeClient.userId!)")
        DispatchQueue.main.async {
            InstanceManager.shared.viewController?.title = "Connected as \(stringeeClient.userId ?? "")"
        }
        
        CallManager.shared.startCheckingReceivingTimeoutOfStringeeCall()
    }
    
    func didDisConnect(_ stringeeClient: StringeeClient!, isReconnecting: Bool) {
        print("didDisConnect \(stringeeClient.userId!)")
        DispatchQueue.main.async {
            InstanceManager.shared.viewController?.title = "Disconnected"
        }
        CallManager.shared.stopCheckingReceivingTimeoutOfStringeeCall()
    }
    
    func didFailWithError(_ stringeeClient: StringeeClient!, code: Int32, message: String!) {
        print("didFailWithError")
    }
    
    func requestAccessToken(_ stringeeClient: StringeeClient!) {
        print("requestAccessToken")
        stringeeClient.connect(withAccessToken: token)
    }
    
    func didReceiveCustomMessage(_ stringeeClient: StringeeClient!, message: [AnyHashable : Any]!, fromUserId userId: String!) {
        print("didReceiveCustomMessage")
    }
}

extension StringeeImplement: StringeeIncomingCallDelegate {
    func incomingCall(with stringeeClient: StringeeClient!, stringeeCall: StringeeCall!) {
        CallManager.shared.handleIncomingCallEvent(stringeeCall: stringeeCall)
    }
}
